lib/internal/modules/esm/loader.js

```diff
@@ -373,22 +390,44 @@ class ESMLoader {
       const { compileFunction } = require('vm');
       const preloadInit = compileFunction(
         preload,
-        ['getBuiltin'],
+        ['getBuiltin', 'port', 'setImportMetaCallback'],
         {
           filename: '<preload>',
         }
       );
       const { NativeModule } = require('internal/bootstrap/loaders');
-
-      FunctionPrototypeCall(preloadInit, globalThis, (builtinName) => {
-        if (NativeModule.canBeRequiredByUsers(builtinName)) {
-          return require(builtinName);
+      let finished = false;
+      let replacedImportMetaInitializer = false;
+      let next = this.#importMetaInitializer;
+      try {
+        FunctionPrototypeCall(preloadInit, globalThis, (builtinName) => {
+          if (NativeModule.canBeRequiredByUsers(builtinName)) {
+            return require(builtinName);
+          }
+          throw new ERR_INVALID_ARG_VALUE('builtinName', builtinName);
+        }, insidePreload, (fn) => {
+          if (finished || typeof fn !== 'function') {
+            throw new ERR_INVALID_ARG_TYPE('fn', fn);
+          }
+          replacedImportMetaInitializer = true;
+          const parent = next;
+          next = (meta, context) => {
+            return fn(meta, context, parent);
+          };
+        });
+      } finally {
+        finished = true;
+        if (replacedImportMetaInitializer) {
+          this.#importMetaInitializer = next;
```

Comment:
I think this should be punted since moving these out of this file and passing a param instead of `this` doesn't seem clear to me on if we need to have the same abstraction for each hook doing this kind of workflow refactoring (see .#resolvers .#loaders, .#globalPreloaders).